src/script/E2EIdentity/E2EIdentityEnrollment.ts

```diff
@@ -92,32 +103,140 @@ export class E2EIHandler extends TypedEventEmitter<Events> {
+    const renewalPromptTime = new Date(certificateCreationTime + renewalTimeMS).getTime();
+    const currentTime = new Date().getTime();
+
+    // Check if it's time to renew the certificate
+    if (currentTime >= renewalPromptTime) {
+      await this.renewCertificate();
+    }
+  }
+
+  /**
+   * Renew the certificate without user action
+   */
+  private async renewCertificate(): Promise<void> {
+    this.oidcService = new OIDCService();
+    try {
+      // Use the oidc service to get the user data via silent authentication (refresh token)
+      const userData = await this.oidcService.handleSilentAuthentication();
+
+      if (!userData) {
+        throw new Error('Received no user data from OIDC service');
+      }
+      // renew without user action
+      await this.enroll(true, userData);
+    } catch (error) {
+      this.logger.error('Silent authentication with refresh token failed', error);
+
+      // If the silent authentication fails, clear the oidc service progress/data and renew manually
+      await this.cleanUp();
+      this.showE2EINotificationMessage(ModalType.CERTIFICATE_RENEWAL);
+    }
+  }
+
+  /**
+   * Calculates the date when the E2EI certificate renewal should be prompted.
+   *
+   * @param {number} timeRemainingMS - Certificate validity period in days (VP).
+   * @param {number} historyTime - Maximum time messages are stored in days (HT).
+   * @param {number} gracePeriod - Time to activate certificate in days (GP).
+   * @returns {Date} - The date to start prompting for certificate renewal.
+   */
+  private calculateRenewalTime(timeRemainingMS: number, historyTimeMS: number, gracePeriodMS: number) {
+    // Calculate a random time between 0 and 1 days
+    const randomTimeInMS = Math.random() * TimeInMillis.DAY; // Up to 24 hours in milliseconds
+
+    // Calculate the total days to subtract
+    const totalDaysToSubtract = historyTimeMS + gracePeriodMS + randomTimeInMS;
+
+    // Calculate the renewal date
+    const renewalDate = timeRemainingMS - totalDaysToSubtract;
+
+    return renewalDate;
+  }
+
+  get isE2EIEnabled(): boolean {
+    return supportsMLS() && Config.getConfig().FEATURE.ENABLE_E2EI;
+  }
+
   private async storeRedirectTargetAndRedirect(targetURL: string): Promise<void> {
     // store the target url in the persistent oidc service store, since the oidc service will be destroyed after the redirect
     OIDCServiceStore.store.targetURL(targetURL);
     this.oidcService = new OIDCService();
     await this.oidcService.authenticate();
   }
 
-  public async enroll(refreshActiveCertificate: boolean = false) {
+  /**
+   * Used to clean the state/storage after a failed run
+   */
+  private async cleanUp() {
+    // Remove the url parameters of the failed enrolment
+    removeUrlParameters();
+    // Clear the oidc service progress
+    this.oidcService = new OIDCService();
+    await this.oidcService.clearProgress();
```

Comment:
should those 2 lines be in reversed order? It feels we are clearing the progress of a completely new instance

Reply:
The progress is stored in localStorage, not depending on the instance. We just make sure that we have a oidcService here and can call clearProgress.